spec.html

```diff
@@ -648,28 +651,46 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
+                1. If _m_.[[Async]] is *false*, perform ! GatherAsyncParentCompletions(_m_, _execList_).
+        1. Return *undefined*.
+      </emu-alg>
+      <emu-note>
+        <p>When an async execution for a root _module_ is fulfilled, this function determines the list of modules which are able to synchronously execute together on this completion, populating them in _execList_.</p>
+      </emu-note>
+    </emu-clause>
+
     <emu-clause id="sec-async-module-execution-fulfilled" aoid="AsyncModuleExecutionFulfilled">
       <h1><ins>AsyncModuleExecutionFulfilled ( _module_ )</ins></h1>
       <emu-alg>
-        1. Assert: _module_.[[Status]] is ~evaluated~.
-        1. If _module_.[[AsyncEvaluating]] is *false*,
-          1. Assert: _module_.[[EvaluationError]] is not ~empty~.
-          1. Return *undefined*.
+        1. Assert: _module_.[[AsyncEvaluating]] is *true*.
         1. Assert: _module_.[[EvaluationError]] is ~empty~.
         1. Set _module_.[[AsyncEvaluating]] to *false*.
-        1. For each Module _m_ of _module_.[[AsyncParentModules]], do
-          1. Decrement _m_.[[PendingAsyncDependencies]] by 1.
-          1. If _m_.[[PendingAsyncDependencies]] is 0 and _m_.[[EvaluationError]] is ~empty~, then
-            1. Assert: _m_.[[AsyncEvaluating]] is *true*.
-            1. If _m_.[[CycleRoot]].[[EvaluationError]] is not ~empty~, return *undefined*.
-            1. If _m_.[[Async]] is *true*, then
-              1. Perform ! ExecuteAsyncModule(_m_).
-            1. Otherwise,
-              1. Let _result_ be _m_.ExecuteModule().
-              1. If _result_ is a normal completion,
-                1. Perform ! AsyncModuleExecutionFulfilled(_m_).
-              1. Otherwise,
-                1. Perform ! AsyncModuleExecutionRejected(_m_, _result_.[[Value]]).
+        1. Let _execList_ be a new empty List.
+        1. Perform ! GatherAsyncParentCompletions(_module_, _execList_).
+        1. Let _sortedExecList_ be a List of elements that are the elements of _execList_, in the order in which they had their [[AsyncEvaluating]] fields set to *true* in InnerModuleEvaluation.
+        1. Assert: All elements of _sortedExecList_ have their [[AsyncEvaluating]] field set to *true*, [[PendingAsyncDependencies]] field set to 0 and [[EvaluationError]] field set to *undefined*.
+        1. For each Module _m_ of _sortedExecList_, do
+          1. If _m_.[[AsyncEvaluating]] is *false*, then
+            1. Assert: _m_.[[EvaluatingError]] is not ~empty~.
+          1. Otherwise, if _m_.[[Async]] is *true*, then
+            1. Perform ! ExecuteAsyncModule(_m_).
+          1. Otherwise,
+            1. Let _result_ be _m_.ExecuteModule().
+            1. If _result_ is an abrupt completion,
+              1. Perform ! AsyncModuleExecutionRejected(_m_, _result_.[[Value]]).
+            1. Otherwise, set _m_.[[AsyncEvaluating]] to *false*.
         1. If _module_.[[TopLevelCapability]] is not ~empty~, then
```

Comment:
Continuing here, all references to _module_ from this step and below should be replaced with _topLevelParent_.

If an async module didn't get any parents from GatherAsyncParentCompletions, then it itself must be an entry point so resolve its top-level capability. If it did get parents, since it was async, the last parent in _execList_ might be the entry point, so check if it is and resolve the top-level capability.